### PR TITLE
video: wayland: Remove excessive SetFullscreen calls

### PR DESCRIPTION
**Pull request by Kontrabant**
Eliminate excessive calls to `SetFullscreen` by removing the calls in the libdecor and xdg-toplevel config callbacks.

These calls were being made there in case something explicitly called the window minimization function from within SDL, which unsets fullscreen, and as minimizing a window in Wayland is just a suggestion to the compositor and doesn't actually change the window state or communicate anything back to the application, it was necessary to call SetFullscreen in every call to the config functions just in case something minimized a window via `SDL_MinimizeWindow()` and later needed to restore it.  GNOME in particular had issues with strange behavior when fullscreen set/unset operations were being hammered and commits were being performed on surfaces while switching to fullscreen.

With the new video system flag to disable unsetting fullscreen when minimizing a window, these calls in the configuration functions are no longer needed and can be removed. This significantly reduces calls to the `SetFullscreen` function, reverts #6044 for a more proper fix, and fixes a similar problem when hiding and showing a window initially created with fullscreen flags.

Draft for now, as, while this works well from my testing and eliminates a huge amount of fullscreen set/unset calls and commits, I'd like more eyes on it to make sure I'm not missing anything here.

@flibitijibibo 
@dos1

### CODE REVIEW

File: src/video/SDL_video.c [hunk 1]
```diff
@@ -299,6 +299,19 @@ SDL_CreateWindowTexture(SDL_VideoDevice *_this, SDL_Window * window, Uint32 * fo
...
 
+/* Convenience functions for reading driver flags */
+SDL_FORCE_INLINE SDL_bool
+DisableDisplayModeSwitching()
+{
+    return !!(_this->device_flags & DEVICE_FLAG_DISABLE_DISPLAY_MODE_SWITCHING);
+}
+
+SDL_FORCE_INLINE SDL_bool
+DisableUnsetFullscreenOnMinimize()
```
Comment (slouken) [inline src/video/SDL_video.c:310]: These should take _THIS as a parameter rather than reading from the global.

**Comment by dos1** [comment]
Seems reasonable.

> GNOME in particular had issues with strange behavior when fullscreen set/unset operations were being hammered and commits were being performed on surfaces while switching to fullscreen.

If you have a reproducer, that's really something worth reporting to GNOME. Unless SDL was violating the protocol (which from your description sounds like it may not have been the case), the compositor really shouldn't behave in this way.

**Comment by Kontrabant** [comment]
Looking at the Wayland debug logs, I'm not sure there's anything that would be a bug on GNOME's end.  Previously, there appeared to be some cases where additional calls to set the fullscreen state were being made before the previous ones were acked (which seemed to correspond with the glitched fullscreen I was seeing), and GNOME tends to be strict regarding order of operations, even if what the application is doing isn't a protocol violation.

**Comment by dos1** [comment]
If the compositor is more strict than the protocol, that's a compositor bug. Although it sure is inelegant to send unnecessary requests, there's nothing in xdg-shell that would prevent the client from (un)setting fullscreen multiple times - the compositor is free to either send configure requests for each of them or ignore all but the latest one, and as long as the client commits a correct state according to acked configure event (so in case of a fullscreen state, with surface size not larger than requested by configure event) it shouldn't break anything.

**Comment by Kontrabant** [comment]
The glitch where things are offset by the top bar seems to specifically occur when there are overlapping fullscreen calls. When the glitch manifests itself, I see something in the debug logs like this:

- Call to enter fullscreen
- Config from the fullscreen request
- Another call to enter fullscreen
- Ack
- Config from the second fullscreen request
- Ack

Most likely GNOME is seeing the ack intended for the first fullscreen call right after the second fullscreen call and considers it done, while the actual config event for the second fullscreen call is still forthcoming.

It may ultimately be a GNOME issue where the compositor is stricter than the spec, but this seems sloppy on the SDL side as well and cleaning up an excess of fullscreen calls does properly fix the glitches/regressions from introducing the roundtrip in `Wayland_ShowWindow()`while not introducing any side effects (#6044 fixes the most urgent issue of windows created as fullscreen being offset, but if you then call a show/hide on that window, it's offset again).

I'm removing the draft status as I've tested this on GNOME and KDE and I haven't found any regressions.  If possible, it would be good to test it on more niche compositors just to be sure, though.

**Comment by dos1** [comment]
> Most likely GNOME is seeing the ack intended for the first fullscreen call right after the second fullscreen call and considers it done, while the actual config event for the second fullscreen call is still forthcoming.

That would be clearly a compositor bug then, since acknowledgements have serial numbers that tie them to specific configure events.

> but this seems sloppy on the SDL side as well

No disagreement there at all :) But could you fill a Mutter issue for that anyway? That's not how the compositor should behave and it may still end up biting us or someone else in the future.

**Pull request merged**